jupyterhub_traefik_proxy/toml.py

```diff
@@ -0,0 +1,300 @@
+
+Route Specification:
+
+- A routespec is a URL prefix ([host]/path/), e.g.
+  'host.tld/path/' for host-based routing or '/path/' for default routing.
+- Route paths should be normalized to always start and end with '/'
+"""
+
+# Copyright (c) Jupyter Development Team.
+# Distributed under the terms of the Modified BSD License.
+
+import json
+import os
+import asyncio
+
+from urllib.parse import urlparse
+from traitlets import Any, default, Unicode
+
+from . import traefik_utils
+from jupyterhub.proxy import Proxy
+from jupyterhub_traefik_proxy import TraefikProxy
+
+
+class TraefikTomlProxy(TraefikProxy):
+    """JupyterHub Proxy implementation using traefik and toml config file"""
+
+    mutex = Any()
+
+    @default("mutex")
+    def _default_mutex(self):
+        return asyncio.Lock()
+
+    toml_static_config_file = Unicode(
+        "traefik.toml", config=True, help="""traefik's static configuration file"""
+    )
+
+    toml_dynamic_config_file = Unicode(
+        "rules.toml", config=True, help="""traefik's dynamic configuration file"""
+    )
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+        self.routes_cache = {}
+
+    async def _setup_traefik_static_config(self):
+        self.log.info("Setting up traefik's static config...")
+        self._generate_htpassword()
+
+        try:
+            with open(self.toml_static_config_file, "w") as f:
+                f.write('defaultEntryPoints = ["http"]\n')
+                f.write("debug = true\n"),
+                f.write('logLevel = "ERROR"\n')
+
+                f.write("[entryPoints]\n"),
+                f.write("\t[entryPoints.http]\n")
+                f.write(
+                    '\t\taddress = ":' + str(urlparse(self.public_url).port) + '"\n'
+                )
+                f.write("\t[entryPoints.auth_api]\n")
+                f.write(
+                    '\t\taddress = ":'
+                    + str(urlparse(self.traefik_api_url).port)
+                    + '"\n'
+                )
+                f.write("\t\t[entryPoints.auth_api.auth]\n")
+                f.write("\t\t\t[entryPoints.auth_api.auth.basic]\n")
+                f.write(
+                    '\t\t\t\tusers = ["'
+                    + self.traefik_api_username
+                    + ":"
+                    + self.traefik_api_hashed_password
+                    + '"]\n'
+                )
+
+                f.write("[api]\n"),
+                f.write("\tdashboard = true\n"),
+                f.write('\tentrypoint = "auth_api"\n'),
+
+                f.write("[file]\n")
+                f.write('\tfilename = "' + self.toml_dynamic_config_file + '"\n')
+                f.write("\twatch = true\n")
+            open(self.toml_dynamic_config_file, "a").close()
+        except IOError as e:
+            self.log.error(
+                "Couldn't set up traefik's static config. I/O error({0}): {1}".format(
+                    e.errno, e.strerrornse
```

Comment:
If we use `self.log.exception("msg...")`, the logger will take care of formatting the exception itself, including a traceback.